File: backend/services/sanitize.py
"""
Input sanitization utilities for XSS and injection prevention
TC-87: Güvenlik & Spam Önlemi
"""

import re
import html
from typing import Dict, List, Tuple, Optional


# HTML entities that need escaping
HTML_ESCAPE_TABLE = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&#x27;",
    "/": "&#x2F;",
    "`": "&#x60;",
    "=": "&#x3D;",
}


def escape_html(text: str) -> str:
    """
    Escape HTML special characters to prevent XSS attacks.
    
    Args:
        text: The input string to sanitize
        
    Returns:
        Sanitized string with HTML entities escaped
    """
    if not text or not isinstance(text, str):
        return ""
    
    return html.escape(text, quote=True)


def strip_dangerous_tags(text: str) -> str:
    """
    Remove potentially dangerous HTML tags and scripts.
    
    Args:
        text: The input string to sanitize
        
    Returns:
        Sanitized string with dangerous content removed
    """
    if not text or not isinstance(text, str):
        return ""
    
    # Remove script tags and their content
    text = re.sub(r'<script\b[^<]*(?:(?!</script>)<[^<]*)*</script>', '', text, flags=re.IGNORECASE)
    
    # Remove event handlers (onclick, onerror, onload, etc.)
    text = re.sub(r'\s*on\w+\s*=\s*["\'][^"\']*["\']', '', text, flags=re.IGNORECASE)
    text = re.sub(r'\s*on\w+\s*=\s*[^\s>]*', '', text, flags=re.IGNORECASE)
    
    # Remove javascript: protocol
    text = re.sub(r'javascript\s*:', '', text, flags=re.IGNORECASE)
    
    # Remove data: protocol for potential XSS
    text = re.sub(r'data\s*:\s*text/html', '', text, flags=re.IGNORECASE)
    
    # Remove iframe, object, embed tags
    text = re.sub(r'<(iframe|object|embed|form)[^>]*>.*?</\1>', '', text, flags=re.IGNORECASE | re.DOTALL)
    text = re.sub(r'<(iframe|object|embed|form)[^>]*/?\s*>', '', text, flags=re.IGNORECASE)
    
    # Remove style tags with expressions
    text = re.sub(r'<style[^>]*>.*?</style>', '', text, flags=re.IGNORECASE | re.DOTALL)
    
    # Remove meta, link, base tags
    text = re.sub(r'<(meta|link|base)[^>]*/?\s*>', '', text, flags=re.IGNORECASE)
    
    return text


# SQL injection patterns to detect and block
SQL_INJECTION_PATTERNS = [
    re.compile(r'\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION|TRUNCATE|GRANT|REVOKE)\b', re.IGNORECASE),
    re.compile(r'(--)|(;)|(\/\*)|(\*\/)'),
    re.compile(r"'(\s*)(OR|AND)(\s*)'?(\s*)(\d+|')", re.IGNORECASE),
    re.compile(r'"(\s*)(OR|AND)(\s*)"?(\s*)(\d+|")', re.IGNORECASE),
    re.compile(r'\b(OR|AND)\b\s+\d+\s*=\s*\d+', re.IGNORECASE),
    re.compile(r"\b(OR|AND)\b\s+'[^']*'\s*=\s*'[^']*'", re.IGNORECASE),
]


def detect_sql_injection(text: str) -> Tuple[bool, List[str]]:
    """
    Check if input contains potential SQL injection patterns.
    
    Args:
        text: The input string to check
        
    Returns:
        Tuple of (is_valid, detected_patterns)
    """
    if not text or not isinstance(text, str):
        return True, []
    
    detected_patterns = []
    
    for pattern in SQL_INJECTION_PATTERNS:
        matches = pattern.findall(text)
        if matches:
            detected_patterns.extend([str(m) if isinstance(m, str) else str(m[0]) for m in matches])
    
    return len(detected_patterns) == 0, detected_patterns


# XSS patterns to detect
XSS_PATTERNS = [
    re.compile(r'<script\b[^<]*(?:(?!</script>)<[^<]*)*</script>', re.IGNORECASE),
    re.compile(r'javascript\s*:', re.IGNORECASE),
    re.compile(r'on\w+\s*=', re.IGNORECASE),
    re.compile(r'<iframe', re.IGNORECASE),
    re.compile(r'<object', re.IGNORECASE),
    re.compile(r'<embed', re.IGNORECASE),
    re.compile(r'<svg\s+onload', re.IGNORECASE),
    re.compile(r'<img\s+[^>]*onerror', re.IGNORECASE),
    re.compile(r'expression\s*\(', re.IGNORECASE),
    re.compile(r'url\s*\(\s*["\']?\s*javascript', re.IGNORECASE),
]


def detect_xss(text: str) -> Tuple[bool, List[str]]:
    """
    Check if input contains potential XSS patterns.
    
    Args:
        text: The input string to check
        
    Returns:
        Tuple of (is_valid, detected_patterns)
    """
    if not text or not isinstance(text, str):
        return True, []
    
    detected_patterns = []
    
    for pattern in XSS_PATTERNS:
        matches = pattern.findall(text)
        if matches:
            detected_patterns.extend(matches)
    
    return len(detected_patterns) == 0, detected_patterns


def sanitize_input(text: str) -> str:
    """
    Comprehensive input sanitization that handles XSS and common injection patterns.
    
    Args:
        text: The input string to sanitize
        
    Returns:
        Sanitized string safe for display and storage
    """
    if not text or not isinstance(text, str):
        return ""
    
    # Step 1: Trim whitespace
    sanitized = text.strip()
    
    # Step 2: Remove dangerous HTML tags and scripts
    sanitized = strip_dangerous_tags(sanitized)
    
    # Step 3: Escape remaining HTML entities
    sanitized = escape_html(sanitized)
    
    return sanitized


def validate_and_sanitize(
    text: str,
    max_length: int = 10000,
    min_length: int = 0,
    allow_html: bool = False,
    check_sql_injection: bool = True,
    check_xss: bool = True,
) -> Dict:
    """
    Validate and sanitize user input for forms.
    
    Args:
        text: The input string to validate and sanitize
        max_length: Maximum allowed length
        min_length: Minimum required length
        allow_html: Whether to allow HTML content
        check_sql_injection: Whether to check for SQL injection
        check_xss: Whether to check for XSS
        
    Returns:
        Dictionary with validation result
    """
    errors = []
    warnings = []
    
    if not text or not isinstance(text, str):
        return {
            "is_valid": min_length == 0,
            "sanitized_value": "",
            "errors": ["Input is required"] if min_length > 0 else [],
            "warnings": [],
        }
    
    # Check length constraints
    if len(text) < min_length:
        errors.append(f"Input must be at least {min_length} characters")
    
    if len(text) > max_length:
        errors.append(f"Input must be no more than {max_length} characters")
    
    # Check for XSS patterns
    if check_xss:
        xss_valid, xss_patterns = detect_xss(text)
        if not xss_valid:
            warnings.append("Potentially unsafe content detected and removed")
    
    # Check for SQL injection patterns
    if check_sql_injection:
        sql_valid, sql_patterns = detect_sql_injection(text)
        if not sql_valid:
            warnings.append("Potentially unsafe patterns detected")
    
    # Sanitize the input
    sanitized_value = text
    if not allow_html:
        sanitized_value = sanitize_input(text)
    
    # Enforce max length on sanitized value
    if len(sanitized_value) > max_length:
        sanitized_value = sanitized_value[:max_length]
    
    return {
        "is_valid": len(errors) == 0,
        "sanitized_value": sanitized_value,
        "errors": errors,
        "warnings": warnings,
    }


def safe_text(text: str) -> str:
    """
    Safe text display - use when rendering user-generated content.
    
    Args:
        text: The text to make safe for display
        
    Returns:
        Safe text for rendering
    """
    return escape_html(text or "")

File: frontend/src/utils/sanitize.ts
/**
 * Input sanitization utilities for XSS and injection prevention
 * TC-87: Güvenlik & Spam Önlemi
 */

/**
 * HTML entities map for escaping dangerous characters
 */
const HTML_ENTITIES: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#x27;',
  '/': '&#x2F;',
  '`': '&#x60;',
  '=': '&#x3D;',
};

/**
 * Escape HTML special characters to prevent XSS attacks
 * @param input - The input string to sanitize
 * @returns Sanitized string with HTML entities escaped
 */
export function escapeHtml(input: string): string {
  if (!input || typeof input !== 'string') {
    return '';
  }
  return input.replace(/[&<>"'`=/]/g, (char) => HTML_ENTITIES[char] || char);
}

/**
 * Remove potentially dangerous HTML tags and scripts
 * @param input - The input string to sanitize
 * @returns Sanitized string with dangerous content removed
 */
export function stripDangerousTags(input: string): string {
  if (!input || typeof input !== 'string') {
    return '';
  }
  
  // Remove script tags and their content
  let sanitized = input.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '');
  
  // Remove event handlers (onclick, onerror, onload, etc.)
  sanitized = sanitized.replace(/\s*on\w+\s*=\s*["'][^"']*["']/gi, '');
  sanitized = sanitized.replace(/\s*on\w+\s*=\s*[^\s>]*/gi, '');
  
  // Remove javascript: protocol
  sanitized = sanitized.replace(/javascript\s*:/gi, '');
  
  // Remove data: protocol for potential XSS
  sanitized = sanitized.replace(/data\s*:\s*text\/html/gi, '');
  
  // Remove iframe, object, embed tags
  sanitized = sanitized.replace(/<(iframe|object|embed|form)[^>]*>.*?<\/\1>/gi, '');
  sanitized = sanitized.replace(/<(iframe|object|embed|form)[^>]*\/?>/gi, '');
  
  // Remove style tags with expressions
  sanitized = sanitized.replace(/<style[^>]*>.*?<\/style>/gi, '');
  
  // Remove meta, link, base tags
  sanitized = sanitized.replace(/<(meta|link|base)[^>]*\/?>/gi, '');
  
  return sanitized;
}

/**
 * SQL injection patterns to detect and block
 */
const SQL_INJECTION_PATTERNS = [
  /(\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE|UNION|TRUNCATE|GRANT|REVOKE)\b)/gi,
  /(--)|(;)|(\/\*)|(\*\/)/g,
  /'(\s*)(OR|AND)(\s*)'?(\s*)(\d+|')/gi,
  /"(\s*)(OR|AND)(\s*)"?(\s*)(\d+|")/gi,
  /\b(OR|AND)\b\s+\d+\s*=\s*\d+/gi,
  /\b(OR|AND)\b\s+'[^']*'\s*=\s*'[^']*'/gi,
];

/**
 * Check if input contains potential SQL injection patterns
 * @param input - The input string to check
 * @returns Object with isValid boolean and detected patterns
 */
export function detectSqlInjection(input: string): { isValid: boolean; detectedPatterns: string[] } {
  if (!input || typeof input !== 'string') {
    return { isValid: true, detectedPatterns: [] };
  }
  
  const detectedPatterns: string[] = [];
  
  for (const pattern of SQL_INJECTION_PATTERNS) {
    const matches = input.match(pattern);
    if (matches) {
      detectedPatterns.push(...matches);
    }
  }
  
  return {
    isValid: detectedPatterns.length === 0,
    detectedPatterns,
  };
}

/**
 * XSS patterns to detect
 */
const XSS_PATTERNS = [
  /<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi,
  /javascript\s*:/gi,
  /on\w+\s*=/gi,
  /<iframe/gi,
  /<object/gi,
  /<embed/gi,
  /<svg\s+onload/gi,
  /<img\s+[^>]*onerror/gi,
  /expression\s*\(/gi,
  /url\s*\(\s*["']?\s*javascript/gi,
];

/**
 * Check if input contains potential XSS patterns
 * @param input - The input string to check
 * @returns Object with isValid boolean and detected patterns
 */
export function detectXss(input: string): { isValid: boolean; detectedPatterns: string[] } {
  if (!input || typeof input !== 'string') {
    return { isValid: true, detectedPatterns: [] };
  }
  
  const detectedPatterns: string[] = [];
  
  for (const pattern of XSS_PATTERNS) {
    const matches = input.match(pattern);
    if (matches) {
      detectedPatterns.push(...matches);
    }
  }
  
  return {
    isValid: detectedPatterns.length === 0,
    detectedPatterns,
  };
}

/**
 * Comprehensive input sanitization that handles XSS and common injection patterns
 * @param input - The input string to sanitize
 * @returns Sanitized string safe for display and storage
 */
export function sanitizeInput(input: string): string {
  if (!input || typeof input !== 'string') {
    return '';
  }
  
  // Step 1: Trim whitespace
  let sanitized = input.trim();
  
  // Step 2: Remove dangerous HTML tags and scripts
  sanitized = stripDangerousTags(sanitized);
  
  // Step 3: Escape remaining HTML entities
  sanitized = escapeHtml(sanitized);
  
  return sanitized;
}

/**
 * Validate and sanitize user input for forms
 * Returns validation result and sanitized value
 * @param input - The input string to validate and sanitize
 * @param options - Validation options
 * @returns Validation result with sanitized value
 */
export function validateAndSanitize(
  input: string,
  options: {
    maxLength?: number;
    minLength?: number;
    allowHtml?: boolean;
    checkSqlInjection?: boolean;
    checkXss?: boolean;
  } = {}
): {
  isValid: boolean;
  sanitizedValue: string;
  errors: string[];
  warnings: string[];
} {
  const {
    maxLength = 10000,
    minLength = 0,
    allowHtml = false,
    checkSqlInjection = true,
    checkXss = true,
  } = options;
  
  const errors: string[] = [];
  const warnings: string[] = [];
  
  if (!input || typeof input !== 'string') {
    return {
      isValid: minLength === 0,
      sanitizedValue: '',
      errors: minLength > 0 ? ['Input is required'] : [],
      warnings: [],
    };
  }
  
  // Check length constraints
  if (input.length < minLength) {
    errors.push(`Input must be at least ${minLength} characters`);
  }
  
  if (input.length > maxLength) {
    errors.push(`Input must be no more than ${maxLength} characters`);
  }
  
  // Check for XSS patterns
  if (checkXss) {
    const xssResult = detectXss(input);
    if (!xssResult.isValid) {
      warnings.push('Potentially unsafe content detected and removed');
    }
  }
  
  // Check for SQL injection patterns
  if (checkSqlInjection) {
    const sqlResult = detectSqlInjection(input);
    if (!sqlResult.isValid) {
      warnings.push('Potentially unsafe patterns detected');
    }
  }
  
  // Sanitize the input
  let sanitizedValue = input;
  if (!allowHtml) {
    sanitizedValue = sanitizeInput(input);
  }
  
  // Enforce max length on sanitized value
  if (sanitizedValue.length > maxLength) {
    sanitizedValue = sanitizedValue.substring(0, maxLength);
  }
  
  return {
    isValid: errors.length === 0,
    sanitizedValue,
    errors,
    warnings,
  };
}

/**
 * Safe text display - use when rendering user-generated content
 * @param text - The text to make safe for display
 * @returns Safe text for rendering
 */
export function safeText(text: string): string {
  return escapeHtml(text || '');
}
